pkg/workloadlearning/cache.go

```diff
@@ -0,0 +1,148 @@
+	se, err := cw.sysSessionPool.Get()
+	if err != nil {
+		logutil.BgLogger().Warn("Get system session failed when updating table cost cache", zap.Error(err))
+		return
+	}
+	defer func() {
+		if err == nil { // only recycle when no error
+			cw.sysSessionPool.Put(se)
+		} else {
+			// Note: Otherwise, the session will be leaked.
+			cw.sysSessionPool.Destroy(se)
+		}
+	}()
+
+	sctx := se.(sessionctx.Context)
+	exec := sctx.GetRestrictedSQLExecutor()
+	ctx := kv.WithInternalSourceType(context.Background(), kv.InternalTxnWorkloadLearning)
+
+	// Whether to update table cost metrics
+	// Get the latest latestVersionInStorage in the storage
+	sql := `SELECT version FROM mysql.tidb_workload_values
+            WHERE category = %? AND type = %?
+            ORDER BY version DESC LIMIT 1`
+	rows, _, err := exec.ExecRestrictedSQL(ctx, nil, sql, feedbackCategory, tableReadCost)
+	if err != nil {
+		logutil.BgLogger().Warn("Failed to get the latest table cost version", zap.Error(err))
+		return
+	}
+	// Case: no metrics belongs to this feedback category and type
+	if len(rows) != 1 {
+		logutil.BgLogger().Warn("The result of latest table cost version query is not 1",
+			zap.Int("result_rows", len(rows)),
+			zap.Error(err))
+		return
+	}
+	// If the latest latestVersionInStorage is the same as the cached latestVersionInStorage, no need to update
+	latestVersionInStorage := rows[0].GetUint64(0)
+	if latestVersionInStorage <= cw.tableReadCostCache.Version {
+		logutil.BgLogger().Info("The latest table cost version in storage is the same as the cached version, no need to update",
+			zap.Uint64("latest_version_in_storage", latestVersionInStorage),
+			zap.Uint64("cached_version", cw.tableReadCostCache.Version))
+		return
+	}
+
+	// Get the latest table cost of metrics
+	sql = `SELECT table_id, value FROM mysql.tidb_workload_values
+            WHERE category = %? AND type = %? AND version = %?`
+	rows, _, err = exec.ExecRestrictedSQL(ctx, nil, sql, feedbackCategory, tableReadCost, latestVersionInStorage)
+	if err != nil {
+		logutil.BgLogger().Warn("Failed to get the latest table cost metrics",
+			zap.Error(err))
+		return
+	}
+	newMetrics := make(map[int64]*TableReadCostMetrics)
+	for _, row := range rows {
+		tableID := row.GetInt64(0)
+		value := row.GetBytes(1)
+
+		metric := &TableReadCostMetrics{}
+		if err := json.Unmarshal(value, metric); err != nil {
+			logutil.BgLogger().Warn("Failed to unmarshal table cost metrics",
+				zap.Int64("table_id", tableID),
+				zap.Error(err))
+			continue
+		}
+		newMetrics[tableID] = metric
+	}
+
+	// Update cache atomically
+	cw.updateTableReadCostCacheWithMetrics(newMetrics, latestVersionInStorage)
+}
+
+func (cw *WLCacheWorker) updateTableReadCostCacheWithMetrics(newMetrics map[int64]*TableReadCostMetrics,
+	latestVersionInStorage uint64) {
+	cw.RWMutex.Lock()
+	cw.tableReadCostCache.TableReadCostMetrics = newMetrics
+	cw.tableReadCostCache.Version = latestVersionInStorage
+	cw.RWMutex.Unlock()
```

Comment:
better to use defer for safety (for example, if the above line panic, then we'll hold this lock forever)

Reply:
Done